diff --git a/packages/plugin-pack/sources/commands/pack.ts b/packages/plugin-pack/sources/commands/pack.ts
--- a/packages/plugin-pack/sources/commands/pack.ts
+++ b/packages/plugin-pack/sources/commands/pack.ts
@@ -55,8 +55,12 @@

     if (!workspace)
       throw new WorkspaceRequiredError(project.cwd, this.context.cwd);

+    if (process.env.YARNNIX_PACK_DIRECTORY != null) {
+      workspace.cwd = process.env.YARNNIX_PACK_DIRECTORY
+    }
+
     if (await packUtils.hasPackScripts(workspace)) {
       if (this.installIfNeeded) {
         await project.install({
           cache: await Cache.find(configuration),
